keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/pause/ResourcePauser.kt

```diff
@@ -0,0 +1,84 @@
+/*
+ *
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.netflix.spinnaker.keel.pause
+
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ResourceId
+import com.netflix.spinnaker.keel.api.application
+import com.netflix.spinnaker.keel.api.id
+import com.netflix.spinnaker.keel.events.ResourceActuationPaused
+import com.netflix.spinnaker.keel.events.ResourceActuationResumed
+import com.netflix.spinnaker.keel.persistence.PausedRepository
+import com.netflix.spinnaker.keel.persistence.ResourceRepository
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.PAUSED
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+
+@Component
+class ResourcePauser(
+  val resourceRepository: ResourceRepository,
+  val pausedRepository: PausedRepository,
+  val publisher: ApplicationEventPublisher
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  fun isPaused(resource: Resource<*>): Boolean =
+    pausedRepository.applicationIsPaused(resource.application) || resourceRepository.getStatus(resource.id) == PAUSED
+
+  fun pauseApplication(application: String) {
+    log.info("Pausing application $application")
+    pausedRepository.pauseApplication(application)
+    resourcesInApplication(application).forEach { resource ->
+      publisher.publishEvent(ResourceActuationPaused(resource, "Management of application $application has been paused"))
```

Comment:
When pausing an application (or later, environment) can we avoid updating the history of their child resources? It'd be nice to avoid the multiplicative db writes, but there's also real potential for a user-facing consistency issue. Assume at some point, keel instances will crash in the middle of an app level (un)pause request and we'll get an application paused but only a subset of its resources will land as `ResourceStatus.PAUSED`, or the inverse, an application gets unpaused but an arbitrary number of resources are left in `ResourceStatus.PAUSED`. The user can just retry the app un(pause) but may not notice its required until later, potentially resulting in a confidence hit.

Reply:
Good point. I'll update this.

Reply:
Re: "no status entry in the log if the application was paused", what if `EventController` checks application/env pause status and injects that kind of `ResourceEvent` at the head of event list response? Maybe an enhancement for a future PR.

Reply:
Oh that could be a good idea! Future PR though.

Reply:
https://github.com/spinnaker/keel/issues/648